c2rust-analyze/src/panic_detail.rs

```diff
@@ -0,0 +1,170 @@
+use backtrace::Backtrace;
+use log::warn;
+use rustc_span::{Span, DUMMY_SP};
+use std::any::Any;
+use std::cell::Cell;
+use std::fmt::Write as _;
+use std::panic::PanicInfo;
+
+/// Detailed information about a panic.
+#[derive(Clone, Debug)]
+pub struct PanicDetail {
+    msg: String,
+    loc: Option<String>,
+    relevant_loc: Option<String>,
+    backtrace: Option<Backtrace>,
+    sp: Span,
+}
```

Comment:
Could we separate `PanicDetail` (an actual panic occurred) and `FnFailure` (either a panic occurred or a called function failed)?

Reply:
> Could we separate `PanicDetail` (an actual panic occurred) and `FnFailure` (either a panic occurred or a called function failed)?

I think it's okay to leave this for now, since the "called function failed" will (hopefully) be going away some time soon.

>we could separate the data itself with the formatting/displaying

I'm not entirely clear on what you're suggesting, but this also seems fine to me as-is.  Such implementation details of the `panic_detail` module don't affect much outside the module itself, which is small and seems unlikely to need major changes in the future.

Reply:
Most of these fields are optional, all except for `msg`.  They are all set when an actual panic occurs, but when it's a function failing due to one its callees failing, only `msg` is used.  I was thinking it could like this:

```rust
struct PanicDetail {
    msg: String,
    loc: String,
    backtrace: Backtrace,
    span: Span,
}

impl PanicDetail {
    pub fn from_info(info: &PanicInfo) -> Self {
        Self {
            msg: panic_to_string(info.payload()),
            loc: info.location().unwrap().to_string(),
            backtrace: Backtrace::new_unresolved(),
            span: CURRENT_SPAN.with(|cell| cell.get()),
        }
    }
}

enum FnFailure {
    Panic(PanicDetail),
    CalleeFailed(String),
}
```

Also, as I commented in https://github.com/immunant/c2rust/pull/876#discussion_r1155265976, doing `info.location().unwrap()` simplifies things here more than just that one line, as then uses of `PanicDetail::loc` don't need to handle the `Option`.

I'm not sure what you mean by the "called function failed" will be going away soon, though.

Reply:
>I'm not sure what you mean by the "called function failed" will be going away soon, though.

I'm currently working on better error recovery, so that we'll no longer need to propagate analysis failures up to callers.